pacakges/server/src/core.rs

```diff
@@ -144,85 +157,124 @@ pub fn set_type_and_add_credits(
     bank_name: String,
     sum_credits: &mut f32,
 ) -> bool {
-    if course_status.r#type.is_none()
-        || (course_status.modified && *course_status.r#type.as_ref().unwrap() == bank_name)
-    {
-        course_status.set_type(bank_name);
-        course_status.modified = false; // finished handle this course
-        if course_status.passed() {
-            *sum_credits += course_status.course.credit;
-            return true;
-        }
+    course_status.set_type(bank_name);
+    course_status.modified = false; // finished handle this course
+    if course_status.passed() {
+        *sum_credits += course_status.course.credit;
+        return true;
+    } else {
+        false
     }
-    false
 }
 
 struct BankRuleHandler<'a> {
     user: &'a mut UserDetails,
     bank_name: String,
     course_list: Vec<u32>,
+    courses: &'a HashMap<u32, Course>,
     credit_overflow: f32,
     courses_overflow: Option<u32>,
+    catalog_replacements: HashMap<u32, Replacements>,
+    common_replacements: HashMap<u32, Replacements>,
+    ignore_courses: Vec<u32>
 }
 
 impl<'a> BankRuleHandler<'a> {
-    fn iterate_course_list(&mut self) -> (f32, u32) {
+
+    fn create_optional_courses_list(&self, course_number: u32) -> Replacements {
+        let mut course_list = vec![course_number];
+        if let Some(optional_replacements) = self.catalog_replacements.get(&course_number) {
+            course_list.append(&mut (optional_replacements.clone()));
+        }
+        if let Some(optional_replacements) = self.common_replacements.get(&course_number) {
+            course_list.append(&mut (optional_replacements.clone()));
+        }
+        course_list
+    }
+
+    fn iterate_course_list(&mut self) -> CreditInfo { // return sum_credits, count_courses, missing_points
         let mut sum_credits = self.credit_overflow;
         let mut count_courses = match &self.courses_overflow {
             Some(num_courses) => *num_courses,
             None => 0,
         };
+        let mut missing_credits = 0.0;
         for course_number in &self.course_list {
-            if let Some(course_status) = self.user.get_mut_course_status(*course_number) {
-                let course_added = set_type_and_add_credits(
+            let mut course_added = false;
+            let optional_courses_for_course = self.create_optional_courses_list(*course_number);
```

Comment:
after changing the function's name you could also change "optional_courses_for_course" to something more clear like "optional_course_replacements" or just "course_replacements"